Support free-threaded CPython (3.14t)

This issue is intended to track the efforts of adding support for the free-threaded build of CPython (e.g. 3.14t). The [Python free-threading guide](https://py-free-threading.github.io/porting-extensions/) contains some helpful advice on approaching this task. Based on that guide, I think the following steps make sense:

1. Decide on the thread safety guarantees that apply to the PyTables package
2. Set the Cython `freethreading_compatible` flag
3. Resolve thread-safety issues (mutable global state is likely the most common issue)
4. Add CI testing for free-threaded CPython
5. Provide free-threaded wheels on PyPI (e.g. cp314t).

Regarding step 1 and based on the [threading advice](https://www.pytables.org/cookbook/threading.html) in the cookbook, it seems PyTables already works in a multi-threaded program if some care is taken.  E.g. avoid sharing HDF5 file handles between threads.   The free-threading guide explicitly notes this is valid level of support:
> It is a perfectly valid choice to, for example, only support running on the free-threaded build in effectively single-threaded contexts and not support shared use of objects. It is then up to the users of your library to add locking where appropriate or needed. The advantage of this choice is that it does not force all consumers of your library to pay any cost associated with ensuring thread safety.

Related PRs:
* gh-1270
* gh-1273
* gh-1274

I re-enabled the cp314t wheel builds on my fork of PyTables. Almost everything works, but I'm seeing four test failures running the tests in the windows Wheel build:

https://github.com/ngoldbaum/PyTables/actions/runs/20938117263/job/60165549340?pr=1#step:9:3197

```
..............................................F.F...........F..F..........s..............Exception ignored while finalizing file <tables.nodes.filenode.RAFileNode object at 0x00000327BE886720>:
  Traceback (most recent call last):
    File "C:\Users\runneradmin\AppData\Local\Temp\cibw-run-v_2t0jpd\cp314t-win_amd64\venv-test\Lib\site-packages\tables\nodes\filenode.py", line 142, in close
      if getattr(self._node, "_v_file", None) is None:
  AttributeError: 'RAFileNode' object has no attribute '_node'
  ..Exception ignored while finalizing file <tables.nodes.filenode.RAFileNode object at 0x00000327BE887650>:
  Traceback (most recent call last):
    File "C:\Users\runneradmin\AppData\Local\Temp\cibw-run-v_2t0jpd\cp314t-win_amd64\venv-test\Lib\site-packages\tables\nodes\filenode.py", line 142, in close
      if getattr(self._node, "_v_file", None) is None:
  AttributeError: 'RAFileNode' object has no attribute '_node'
  .................................Exception ignored while finalizing file <tables.nodes.filenode.ROFileNode object at 0x00000327BE88B220>:
  Traceback (most recent call last):
    File "C:\Users\runneradmin\AppData\Local\Temp\cibw-run-v_2t0jpd\cp314t-win_amd64\venv-test\Lib\site-packages\tables\nodes\filenode.py", line 146, in close
      super().close()
    File "C:\Users\runneradmin\AppData\Local\Temp\cibw-run-v_2t0jpd\cp314t-win_amd64\venv-test\Lib\site-packages\tables\nodes\filenode.py", line 157, in flush
      self._check_closed()
    File "C:\Users\runneradmin\AppData\Local\Temp\cibw-run-v_2t0jpd\cp314t-win_amd64\venv-test\Lib\site-packages\tables\nodes\filenode.py", line 365, in _check_closed
      raise ValueError("host PyTables file is already closed!")
  ValueError: host PyTables file is already closed!
  .......
  ======================================================================
  FAIL: None (tables.tests.test_direct_chunk.CArrayDirectChunkingTestCase.None)
  ----------------------------------------------------------------------
  Traceback (most recent call last):
    File "C:\Users\runneradmin\AppData\Local\Temp\cibw-run-v_2t0jpd\cp314t-win_amd64\venv-test\Lib\site-packages\tables\tests\test_direct_chunk.py", line 129, in test_read_chunk
  Performing only a light (yet comprehensive) subset of the test suite.
  If you want a more complete test, try passing the --heavy flag to this script
      self.assertEqual(chunk, obj_bytes)
  (or set the 'heavy' parameter in case you are using tables.test() call).
      ~~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^
  The whole suite will take more than 4 hours to complete on a relatively
  AssertionError: b'x\x01c`dec\x00\x02\x00\x00K\x00\r' != b'x\x01c`dec```\x00\x00\x00K\x00\r'
  modern CPU and around 512 MB of main memory.
  -=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=
  
  ======================================================================
  FAIL: None (tables.tests.test_direct_chunk.CArrayDirectChunkingTestCase.None)
  ----------------------------------------------------------------------
  Traceback (most recent call last):
    File "C:\Users\runneradmin\AppData\Local\Temp\cibw-run-v_2t0jpd\cp314t-win_amd64\venv-test\Lib\site-packages\tables\tests\test_direct_chunk.py", line 145, in test_read_chunk_out
      self.assertRaises(
      ~~~~~~~~~~~~~~~~~^
          ValueError, self.array.read_chunk, chunk_start, out=chunk_out
          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
      )
      ^
  AssertionError: ValueError not raised by read_chunk
  
  ======================================================================
  FAIL: None (tables.tests.test_direct_chunk.EArrayDirectChunkingTestCase.None)
  ----------------------------------------------------------------------
  Traceback (most recent call last):
    File "C:\Users\runneradmin\AppData\Local\Temp\cibw-run-v_2t0jpd\cp314t-win_amd64\venv-test\Lib\site-packages\tables\tests\test_direct_chunk.py", line 129, in test_read_chunk
      self.assertEqual(chunk, obj_bytes)
      ~~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^
  AssertionError: b'x\x01c`dec\x00\x02\x00\x00K\x00\r' != b'x\x01c`dec```\x00\x00\x00K\x00\r'
  
  ======================================================================
  FAIL: None (tables.tests.test_direct_chunk.EArrayDirectChunkingTestCase.None)
  ----------------------------------------------------------------------
  Traceback (most recent call last):
    File "C:\Users\runneradmin\AppData\Local\Temp\cibw-run-v_2t0jpd\cp314t-win_amd64\venv-test\Lib\site-packages\tables\tests\test_direct_chunk.py", line 145, in test_read_chunk_out
      self.assertRaises(
      ~~~~~~~~~~~~~~~~~^
          ValueError, self.array.read_chunk, chunk_start, out=chunk_out
          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
      )
      ^
  AssertionError: ValueError not raised by read_chunk
  
  ----------------------------------------------------------------------
  Ran 6854 tests in 293.834s
  
  FAILED (failures=4, skipped=171)
```

@nascheme any chance you can try to run this down? Maybe something is incorrectly set up in the Windows wheel build.

--- a/pyproject.toml
+++ b/pyproject.toml
@@ -132,9 +132,7 @@ tables = [ "*.dll", "*.so*", "*.dylib" ]
 [tool.cibuildwheel]
 skip = "*-musllinux_*"
 archs = "native"
-# cp314t segfaults, see https://github.com/PyTables/PyTables/issues/1269
-# once it works it should be added here
-build = "cp311-*"
+build = "cp311-* cp314t-*"
 build-verbosity = 1
 test-command = [
     "python -c \"import tables; keys = 'zlib bzip2 blosc blosc2'.split(); missing = [key for key in keys if tables.which_lib_version(key) is None]; assert missing == [], missing\"",
@@ -157,7 +155,8 @@ before-all = [
 ]
 repair-wheel-command = [
     "auditwheel repair -w {dest_dir} {wheel}",
-    "pipx run abi3audit --strict --report {wheel}",
+    # Free-threaded Python 3.14 doesn't support the stable ABI, so skip abi3audit
+    "if [[ {wheel} != *-cp314-cp314t-* ]]; then pipx run abi3audit --strict --report {wheel}; fi",
 ]
 
 [tool.cibuildwheel.linux.environment]
@@ -174,7 +173,8 @@ before-all = [
 repair-wheel-command = [
     "DYLD_LIBRARY_PATH=/tmp/hdf5/lib delocate-listdeps {wheel}",
     "DYLD_LIBRARY_PATH=/tmp/hdf5/lib delocate-wheel --require-archs {delocate_archs} -w {dest_dir} {wheel}",
-    "pipx run abi3audit --strict --report {wheel}",
+    # Free-threaded Python 3.14 doesn't support the stable ABI, so skip abi3audit
+    "if [[ {wheel} != *-cp314-cp314t-* ]]; then pipx run abi3audit --strict --report {wheel}; fi",
 ]
 
 [tool.cibuildwheel.macos.environment]
@@ -196,7 +196,8 @@ before-build = [
 # We check for "libblosc2.dll" so don't let delvewheel mangle it
 repair-wheel-command = [
     "delvewheel repair --no-mangle libblosc2.dll -v -w {dest_dir} {wheel}",
-    "pipx run abi3audit --strict --report {wheel}",
+    # Free-threaded Python 3.14 doesn't support the stable ABI, so skip abi3audit
+    "echo {wheel} | findstr /C:\"cp314t\">nul && call || pipx run abi3audit --strict --report {wheel}",
 ]
 
 [tool.cibuildwheel.windows.environment]

